stdlib/asm/crypto/hashes/blake3.masm

```diff
@@ -0,0 +1,953 @@
+    dup.5
+    u32xor
+    u32rotr.8
+    swap.4
+    drop
+
+    dup.5
+    u32xor
+    u32rotr.8
+
+    swap
+    dup.6
+    u32xor
+    u32rotr.8
+    swap
+
+    dup.2
+    dup.8
+    u32xor
+    u32rotr.8
+    swap.3
+    drop
+
+    # --- #
+
+    pushw.local.0
+    drop
+    drop
+    swap
+    drop
+    pushw.mem # = c #
+
+    dup.2
+    dup.8
+    u32add.unsafe
+    drop
+    swap.3
+    drop
+
+    dup.3
+    dup.5
+    u32add.unsafe
+    drop
+    swap.4
+    drop
+
+    dup.5
+    u32add.unsafe
+    drop
+
+    swap
+    dup.6
+    u32add.unsafe
+    drop
+    swap
+
+    # --- #
+
+    movupw.3
+
+    swap
+    dup.6
+    u32xor
+    u32rotr.7
+    swap
+
+    dup.2
+    dup.8
+    u32xor
+    u32rotr.7
+    swap.3
+    drop
+
+    dup.3
+    dup.5
+    u32xor
+    u32rotr.7
+    swap.4
+    drop
+
+    dup.5
+    u32xor
+    u32rotr.7
+
+    movdnw.3
+
+    # --- #
+
+    pushw.local.0
+    drop
+    drop
+    swap
+    drop
+    popw.mem # = c #
+
+    pushw.local.0
+    drop
+    drop
+    drop
+    popw.mem # = d #
+
+    pushw.local.0
+    repeat.3
+        swap
+        drop
+    end
+    popw.mem # = a #
+
+    pushw.local.0
+    drop
+    repeat.2
+        swap
+        drop
+    end
+    popw.mem # = b #
+end
+
+proc.prepare_columnar_mixing_in_words.0
+    dupw.1
+    dupw.1
+
+    movup.6
+    movup.5
+    movup.4
+    movup.3
+end
+
+proc.prepare_diagonal_mixing_in_words.0
+    dupw.3
+    dupw.3
+
+    movup.6
+    movup.5
+    movup.4
+    movup.3
+end
+
+# see https://github.com/BLAKE3-team/BLAKE3/blob/da4c792/reference_impl/reference_impl.rs#L54-L65 #
+proc.round.1
+    pushw.mem
+    popw.local.0
+
+    # --- columnar mixing --- #
+    # equivalent to https://github.com/BLAKE3-team/BLAKE3/blob/da4c792/reference_impl/reference_impl.rs#L55-L59 #
+    exec.prepare_columnar_mixing_in_words
+    push.env.locaddr.0
+    exec.columnar_mixing
+
+    # --- diagonal mixing --- #
+    # equivalent to https://github.com/BLAKE3-team/BLAKE3/blob/da4c792/reference_impl/reference_impl.rs#L60-L64 #
+    exec.prepare_diagonal_mixing_in_words
+    push.env.locaddr.0
+    exec.diagonal_mixing
+end
+
+# see https://github.com/itzmeanjan/blake3/blob/f07d32e/include/blake3.hpp#L1705-L1759 #
+proc.compress.1
+    popw.local.0
+
+    # round 0 #
+    push.env.locaddr.0
+    exec.round
+    exec.blake3_msg_words_permute
+
+    # round 1 #
+    push.env.locaddr.0
+    exec.round
+    exec.blake3_msg_words_permute
+
+    # round 2 #
+    push.env.locaddr.0
+    exec.round
+    exec.blake3_msg_words_permute
+
+    # round 3 #
+    push.env.locaddr.0
+    exec.round
+    exec.blake3_msg_words_permute
+
+    # round 4 #
+    push.env.locaddr.0
+    exec.round
+    exec.blake3_msg_words_permute
+
+    # round 5 #
+    push.env.locaddr.0
+    exec.round
+    exec.blake3_msg_words_permute
+
+    # round 6 #
+    push.env.locaddr.0
+    exec.round
+    # no permutation required after last round of mixing #
+end
+
+# blake3 2-to-1 hash function #
```

Comment:
Might be good to describe the interface of the procedure in the comments. For example, it expects the values to be hashed to be at the top of the stack (16 elements each containing a u32 value), and the resulting digest (8 u32 values) will be at the top of the stack - or something like this.

Reply:
Done https://github.com/maticnetwork/miden/blob/2f301ab38f1f9baeeedfcde6073f03bd1bf0a154/stdlib/asm/crypto/hashes/blake3.masm#L916-L922